// node-opcua-crypto/dist-esm/source_nodejs/read.js
import assert from "assert";
import fs from "fs";
import path from "path";
import { createPrivateKey, createPublicKey } from "crypto";
import { convertPEMtoDER } from "../source/crypto_utils.js";
var sshpk = require("sshpk");
function _readPemFile(filename) {
    assert(typeof filename === "string");
    return fs.readFileSync(filename, "ascii");
}
function _readPemOrDerFileAsDER(filename) {
    if (filename.match(/.*\.der/)) {
        return fs.readFileSync(filename);
    }
    var raw_key = _readPemFile(filename);
    return convertPEMtoDER(raw_key);
}
export function readCertificate(filename) {
    return _readPemOrDerFileAsDER(filename);
}
export function readPublicKey(filename) {
    if (filename.match(/.*\.der/)) {
        var der = fs.readFileSync(filename);
        return createPublicKey(der);
    }
    else {
        var raw_key = _readPemFile(filename);
        return createPublicKey(raw_key);
    }
}
function myCreatePrivateKey(rawKey) {
    var backup = process.env.OPENSSL_CONF;
    process.env.OPENSSL_CONF = "/dev/null";
    var retValue = createPrivateKey(rawKey);
    process.env.OPENSSL_CONF = backup;
    return retValue;
}
export function readPrivateKey(filename) {
    if (filename.match(/.*\.der/)) {
        var der = fs.readFileSync(filename);
        return myCreatePrivateKey(der);
    }
    else {
        var raw_key = _readPemFile(filename);
        return myCreatePrivateKey(raw_key);
    }
}
export function readCertificatePEM(filename) {
    return _readPemFile(filename);
}
export function readPublicKeyPEM(filename) {
    return _readPemFile(filename);
}
export function readPrivateKeyPEM(filename) {
    return _readPemFile(filename);
}
var __certificate_store = path.join(__dirname, "../../certificates/");
export function setCertificateStore(store) {
    var old_store = __certificate_store;
    __certificate_store = store;
    return old_store;
}
export function readPrivateRsaKey(filename) {
    if (filename.substring(0, 1) !== "." && !fs.existsSync(filename)) {
        filename = __certificate_store + filename;
    }
    var content = fs.readFileSync(filename, "ascii");
    var sshKey = sshpk.parsePrivateKey(content, "auto");
    var key = sshKey.toString("pkcs1");
    return createPrivateKey({ format: "pem", type: "pkcs1", key: key });
}
export function readPublicRsaKey(filename) {
    if (filename.substring(0, 1) !== "." && !fs.existsSync(filename)) {
        filename = __certificate_store + filename;
    }
    var content = fs.readFileSync(filename, "ascii");
    var sshKey = sshpk.parseKey(content, "ssh");
    var key = sshKey.toString("pkcs1");
    return createPublicKey({ format: "pem", type: "pkcs1", key: key });
}
