// node-opcua-crypto/dist-esm/source/crypto_utils2.js
import assert from "assert";
import * as crypto from "crypto";
var createPrivateKey = crypto.createPrivateKey;
import { toPem } from "./crypto_utils.js";
import { coercePrivateKey } from "./x509/coerce_private_key.js";
var jsrsasign = require("jsrsasign");
export function coercePrivateKeyPem(privateKey) {
    if (privateKey instanceof Buffer) {
        var o = createPrivateKey({ key: privateKey, format: "der", type: "pkcs1" });
        var e = o.export({ format: "der", type: "pkcs1" });
        privateKey = toPem(e, "RSA PRIVATE KEY");
    }
    assert(typeof privateKey === "string");
    return privateKey;
}
export function rsaLengthPrivateKey(key) {
    key = coercePrivateKey(key);
    var key2 = key.export({ type: "pkcs1", format: "pem" }).toString();
    var a = jsrsasign.KEYUTIL.getKey(key2);
    return a.n.toString(16).length / 2;
}
export function toPem2(raw_key, pem) {
    assert(raw_key, "expecting a key");
    assert(typeof pem === "string");
    if (raw_key instanceof crypto.KeyObject) {
        if (pem === "RSA PRIVATE KEY") {
            return raw_key.export({ format: "pem", type: "pkcs1" }).toString();
        }
        else if (pem === "PRIVATE KEY") {
            return raw_key.export({ format: "pem", type: "pkcs8" }).toString();
        }
        else {
            throw new Error("Unsupported case!");
        }
    }
    return toPem(raw_key, pem);
}
export function coercePublicKeyPem(publicKey) {
    if (publicKey instanceof crypto.KeyObject) {
        return publicKey.export({ format: "pem", type: "spki" }).toString();
    }
    assert(typeof publicKey === "string");
    return publicKey;
}
export function coerceRsaPublicKeyPem(publicKey) {
    if (publicKey instanceof crypto.KeyObject) {
        return publicKey.export({ format: "pem", type: "spki" }).toString();
    }
    assert(typeof publicKey === "string");
    return publicKey;
}
export function rsaLengthPublicKey(key) {
    key = coercePublicKeyPem(key);
    assert(typeof key === "string");
    var a = jsrsasign.KEYUTIL.getKey(key);
    return a.n.toString(16).length / 2;
}
export function rsaLengthRsaPublicKey(key) {
    key = coerceRsaPublicKeyPem(key);
    assert(typeof key === "string");
    var a = jsrsasign.KEYUTIL.getKey(key);
    return a.n.toString(16).length / 2;
}
